File: macos/Tests/ElevateCoreTests/AccessPackageModelsTests.swift
import Testing
import Foundation
@testable import ElevateCore

@Suite struct AccessPackageModelsTests {
    @Test func requestStateParsesCaseInsensitivelyAndFallsBackToUnknown() {
        #expect(AccessPackageRequestState.parse("PendingApproval") == .pendingApproval)
        #expect(AccessPackageRequestState.parse("delivered") == .delivered)
        #expect(AccessPackageRequestState.parse("partiallyDelivered") == .partiallyDelivered)
        #expect(AccessPackageRequestState.parse("somethingNew") == .unknown)
        #expect(AccessPackageRequestState.parse(nil) == .unknown)
    }

    @Test func requestStateGroupsIntoTabs() {
        #expect(AccessPackageRequestState.pendingApproval.isOpen)
        #expect(AccessPackageRequestState.scheduled.isOpen)
        #expect(!AccessPackageRequestState.delivered.isOpen)
        #expect(AccessPackageRequestState.denied.isDeclined)
        #expect(AccessPackageRequestState.canceled.isDeclined)
        #expect(AccessPackageRequestState.deliveryFailed.isDeclined)
        #expect(!AccessPackageRequestState.submitted.isDeclined)
        #expect(AccessPackageRequestState.submitted.isCancellable)
        #expect(AccessPackageRequestState.pendingApproval.isCancellable)
        #expect(!AccessPackageRequestState.delivering.isCancellable)
    }

    @Test func assignmentStateParses() {
        #expect(AccessPackageAssignmentState.parse("Delivered") == .delivered)
        #expect(AccessPackageAssignmentState.parse("expired") == .expired)
        #expect(AccessPackageAssignmentState.parse("nope") == .unknown)
    }

    @Test func modelsRoundTripThroughCodable() throws {
        let request = AccessPackageRequest(id: "r", packageId: "p", packageName: "Pkg", requestType: "userAdd",
                                           state: .denied, status: "Denied", justification: "why",
                                           createdAt: GraphJSON.parseDate("2026-09-02T09:00:00Z"),
                                           completedAt: GraphJSON.parseDate("2026-09-02T09:03:00Z"), policyId: "pol")
        let assignment = AccessPackageAssignment(id: "a", packageId: "p", packageName: "Pkg", state: .delivered,
                                                 policyName: "All", expiresAt: GraphJSON.parseDate("2027-01-01T00:00:00Z"))
        let snapshot = AccessPackageSnapshot(requests: [request], assignments: [assignment])
        let data = try GraphJSON.encoder.encode(snapshot)
        #expect(try GraphJSON.decoder.decode(AccessPackageSnapshot.self, from: data) == snapshot)
        let requirement = PolicyRequirement(id: "pol", displayName: "Engineers", description: nil, isApprovalRequired: true, requiresAnswers: false)
        #expect(try JSONDecoder().decode(PolicyRequirement.self, from: JSONEncoder().encode(requirement)) == requirement)
    }
}

File: windows/tests/Elevate.Core.Tests/AccessPackageModelsTests.cs
using Elevate.Core.Models;
using Elevate.Core.Tests.Support;
using FluentAssertions;

namespace Elevate.Core.Tests;

/// <summary>Port of the Swift <c>AccessPackageModelsTests</c>.</summary>
public class AccessPackageModelsTests
{
    [Fact]
    public void RequestStateParsesCaseInsensitivelyAndFallsBackToUnknown()
    {
        AccessPackageRequestStates.Parse("PendingApproval").Should().Be(AccessPackageRequestState.PendingApproval);
        AccessPackageRequestStates.Parse("delivered").Should().Be(AccessPackageRequestState.Delivered);
        AccessPackageRequestStates.Parse("partiallyDelivered").Should().Be(AccessPackageRequestState.PartiallyDelivered);
        AccessPackageRequestStates.Parse("somethingNew").Should().Be(AccessPackageRequestState.Unknown);
        AccessPackageRequestStates.Parse("3").Should().Be(AccessPackageRequestState.Unknown);
        AccessPackageRequestStates.Parse(null).Should().Be(AccessPackageRequestState.Unknown);
    }

    [Fact]
    public void RequestStateGroupsIntoTabs()
    {
        AccessPackageRequestState.PendingApproval.IsOpen().Should().BeTrue();
        AccessPackageRequestState.Scheduled.IsOpen().Should().BeTrue();
        AccessPackageRequestState.Delivered.IsOpen().Should().BeFalse();
        AccessPackageRequestState.Denied.IsDeclined().Should().BeTrue();
        AccessPackageRequestState.Canceled.IsDeclined().Should().BeTrue();
        AccessPackageRequestState.DeliveryFailed.IsDeclined().Should().BeTrue();
        AccessPackageRequestState.Submitted.IsDeclined().Should().BeFalse();
        AccessPackageRequestState.Submitted.IsCancellable().Should().BeTrue();
        AccessPackageRequestState.PendingApproval.IsCancellable().Should().BeTrue();
        AccessPackageRequestState.Delivering.IsCancellable().Should().BeFalse();
    }

    [Fact]
    public void AssignmentStateParses()
    {
        AccessPackageAssignmentStates.Parse("Delivered").Should().Be(AccessPackageAssignmentState.Delivered);
        AccessPackageAssignmentStates.Parse("expired").Should().Be(AccessPackageAssignmentState.Expired);
        AccessPackageAssignmentStates.Parse("nope").Should().Be(AccessPackageAssignmentState.Unknown);
        AccessPackageAssignmentStates.Parse(null).Should().Be(AccessPackageAssignmentState.Unknown);
    }

    [Fact]
    public void ModelsRoundTripThroughJsonWithSwiftKeys()
    {
        var request = new AccessPackageRequest("r", "p", "Pkg", "userAdd", AccessPackageRequestState.Denied, "Denied", "why",
            Fixtures.Date("2026-09-02T09:00:00Z"), Fixtures.Date("2026-09-02T09:03:00Z"), "pol");
        var assignment = new AccessPackageAssignment("a", "p", "Pkg", AccessPackageAssignmentState.Delivered, "All", Fixtures.Date("2027-01-01T00:00:00Z"));
        var snapshot = new AccessPackageSnapshot([request], [assignment]);

        var json = Json.Serialize(snapshot);
        json.Should().Contain("\"state\":\"denied\"").And.Contain("\"createdAt\":\"2026-09-02T09:00:00Z\"")
            .And.Contain("\"packageName\":\"Pkg\"").And.Contain("\"expiresAt\":\"2027-01-01T00:00:00Z\"");
        Json.Deserialize<AccessPackageSnapshot>(json).Should().Be(snapshot);

        var requirement = new PolicyRequirement("pol", "Engineers", null, true, false);
        Json.Deserialize<PolicyRequirement>(Json.Serialize(requirement)).Should().Be(requirement);
    }

    [Fact]
    public void SnapshotDefaultsToEmptyListsAndComparesByContent()
    {
        new AccessPackageSnapshot().Requests.Should().BeEmpty();
        new AccessPackageSnapshot().Assignments.Should().BeEmpty();
        new AccessPackageSnapshot().Should().Be(new AccessPackageSnapshot([], []));
        Json.Deserialize<AccessPackageSnapshot>("""{"requests":[],"assignments":[]}""").Should().Be(new AccessPackageSnapshot());
        new AccessPackageSnapshot([new AccessPackageRequest("r", "p", "Pkg", "userAdd", AccessPackageRequestState.Submitted)])
            .Should().NotBe(new AccessPackageSnapshot());
    }
}
